### DISCUSSION TITLE
Isolated Engine

### DISCUSSION THREAD

**Discussion by MaxGrekhov** [sebastienros/jint#1406] [Q&A]
I'm trying to optimize the usage of Engine. And I want to have some sort of isolated engine. This topic has come up a few times already, and I think that the last suggested approach is ShadowRealm. The only problem I see is that It doesn't provide any API to set external value.
It would be nice to have 
```C#
var engine = new Engine(); 
var shadowRealm = engine.Realm.Intrinsics.ShadowRealm.Construct(); 
shadowRealm.SetValue(myContext);
var result = shadowRealm.Evaluate(myScript);
```
Did I miss something? Will it allow to make Engine static or singleton?

**Comment by MaxGrekhov** [comment]
I would like to add that It's not difficult to copy these lines and rework them for ShadowRealm
https://github.com/sebastienros/jint/blob/4839fd19c7829753e753317d213da8d8e77dadc2/Jint/Engine.cs#L177-L216
It works. But I don't know about Shadow Realm architecture and its possible side effects. Any thoughts?
```C#


    public ShadowRealm SetValue(string name, Delegate value)
    {
        _shadowRealm.GlobalObject.FastSetProperty(name, new PropertyDescriptor(new DelegateWrapper(_engine, value), true, false, true));
        return this;
    }

    public ShadowRealm SetValue(string name, string value)
    {
        return SetValue(name, JsString.Create(value));
    }

    public ShadowRealm SetValue(string name, double value)
    {
        return SetValue(name, JsNumber.Create(value));
    }

    public ShadowRealm SetValue(string name, int value)
    {
        return SetValue(name, JsNumber.Create(value));
    }

    public ShadowRealm SetValue(string name, bool value)
    {
        return SetValue(name, value ? JsBoolean.True : JsBoolean.False);
    }

    public ShadowRealm SetValue(string name, JsValue value)
    {
        _shadowRealm.GlobalObject.Set(name, value);
        return this;
    }

    public ShadowRealm SetValue(string name, object obj)
    {
        var value = obj is Type t
            ? TypeReference.CreateTypeReference(_engine, t)
            : JsValue.FromObject(_engine, obj);

        return SetValue(name, value);
    }
```

**Reply by lahma** [reply]
I think such an addition would be beneficial if it still holds the promise of shadow realm protecting original globals from mutation. I'd be happy to review a PR with some added tests cases, there's https://github.com/sebastienros/jint/blob/main/Jint.Tests.PublicInterface/ShadowRealmTests.cs which tries to ensure public API contract and expectations.

**Reply by gentledepp** [reply]
@lahma I created an issue and PR for this.
However, i am facing a problem with the very first unit tests I wrote:
https://github.com/sebastienros/jint/compare/main...gentledepp:jint:feature/1573_shadowrealm?expand=1
```csharp
 [Fact]
    public void CanUseGlobalsDefinedInEngine()
    {
        var engine = new Engine(options => options.EnableModules(GetBasePath()));
        engine.SetValue("message", "world");
        engine.Evaluate("function hello() {return message}");

        Assert.Equal("world",engine.Evaluate("hello();"));

        var shadowRealm = engine.Realm.Intrinsics.ShadowRealm.Construct();
        Assert.Equal("world", shadowRealm.Evaluate("hello();"));
    }
```
I am assuming, that I can 
- create a Jint.Engine and populate its GlobalObject with all the stuff I need.
- can then create a ShadowRealm 
   - and add my own stuff to an isolated GlobalObject
   - and can override stuff of the engines GlobalObject without affecting the engine.
   
   
However, the test fails as the ShadowRealms GlobalObject does not have a function "hello" defined.
I was of the assumption that the "hello" function, as it is defined in the `Engine` should also be available in the `ShadowRealm`....
Am I wrong?

Also: If I add/change something in the engines GlobalObject _after_ the ShadowRealm was created, those changes should _not_ be visible to the ShadowRealm, right? (or should they?)

**Reply by gentledepp** [reply]
I have to apologize - turns out I did not read the docs...
https://github.com/tc39/proposal-shadowrealm/blob/main/explainer.md#how-does-a-shadowrealm-operate
> A newly created shadowRealm does not have immediate access to any object from the incubator realm - and vice-versa - and won't have access to window.top as iframes would.

**Comment by gentledepp** [comment]
One more thing regarding ShadowReadm:
We are caching the parsed script as Esprima.Script and just run it with the Engine.
Would it be feasible to also add an `ShadowRealm.Evaluate(Esprima.Script script)` overload?

**Comment by gentledepp** [comment]
@lahma @MaxGrekhov 
I finally upadted my PR based on yours @MaxGrekhov 

Any thoughts?

https://github.com/sebastienros/jint/issues/1573
